File: rlm/server/heartbeat.py
"""
RLM Heartbeat — Feedback visual generalizado para todos os canais.

Problema real:
    O Telegram gateway já tem um padrão de "keep typing" (thread daemon
    que envia sendChatAction a cada 4s). Mas esse padrão está hardcoded
    dentro de process_and_reply(). WhatsApp, Discord e Slack não têm
    feedback nenhum durante processamento.

Solução:
    Heartbeat generaliza o padrão typing thread do Telegram:
    - Aceita qualquer callable como "ação de heartbeat"
    - Funciona em sync (threading.Event) e async (asyncio.Event)
    - IDisposable — para automaticamente no dispose()
    - CancellationToken — para se o processamento for cancelado

Design RLM-nativo:
    - SyncHeartbeat: thread daemon (Telegram, chamadas sync)
    - AsyncHeartbeat: asyncio.Task (FastAPI gateways)
    - Ambos implementam IDisposable para integração com DisposableStore

Uso:
    # Sync (Telegram)
    hb = SyncHeartbeat(
        action=lambda: _send_typing(token, chat_id),
        interval_s=4.0,
    )
    hb.start()
    try:
        response = rlm.completion(text)
    finally:
        hb.dispose()

    # Async (WhatsApp/Discord/Slack)
    async with AsyncHeartbeat(
        action=lambda: send_typing_indicator(wa_id),
        interval_s=5.0,
    ) as hb:
        response = await run_rlm(text)
"""
from __future__ import annotations

import asyncio
import threading
from typing import Callable, Any

from rlm.core.lifecycle.cancellation import CancellationToken
from rlm.logging import get_runtime_logger

log = get_runtime_logger("heartbeat")


class SyncHeartbeat:
    """
    Heartbeat síncrono via thread daemon.

    Chama `action()` a cada `interval_s` segundos até stop()/dispose().
    Usado pelo Telegram gateway e qualquer processamento em thread.
    """

    __slots__ = (
        "_action", "_interval", "_cancel_token",
        "_stop_event", "_thread", "_disposed", "_unsub",
    )

    def __init__(
        self,
        action: Callable[[], Any],
        interval_s: float = 4.0,
        cancel_token: CancellationToken | None = None,
    ) -> None:
        self._action = action
        self._interval = interval_s
        self._cancel_token = cancel_token
        self._stop_event = threading.Event()
        self._thread: threading.Thread | None = None
        self._disposed = False
        self._unsub: Any = None

    def start(self) -> None:
        """Inicia a thread de heartbeat. Idempotente."""
        if self._thread is not None or self._disposed:
            return

        # Se já cancelado, nem começa
        if self._cancel_token is not None and self._cancel_token.is_cancelled:
            return

        # Registrar callback de cancelamento
        if self._cancel_token is not None:
            self._unsub = self._cancel_token.on_cancelled(self._stop_event.set)

        self._thread = threading.Thread(
            target=self._loop, daemon=True, name="rlm-heartbeat",
        )
        self._thread.start()

    def stop(self) -> None:
        """Para o heartbeat. Thread-safe."""
        self._stop_event.set()

    def dispose(self) -> None:
        """IDisposable — para e limpa."""
        if self._disposed:
            return
        self._disposed = True
        self._stop_event.set()
        if self._unsub is not None:
            self._unsub.dispose()
            self._unsub = None
        # Não faz join() — é daemon thread, não bloqueia shutdown

    def _loop(self) -> None:
        """Loop interno da thread."""
        while not self._stop_event.is_set():
            try:
                self._action()
            except Exception as exc:
                log.debug("Heartbeat action falhou", error=str(exc))
            self._stop_event.wait(timeout=self._interval)


class AsyncHeartbeat:
    """
    Heartbeat assíncrono via asyncio.Task.

    Chama `action()` a cada `interval_s` segundos.
    Pode ser sync callable (chamado inline) ou async callable (awaited).

    Uso como context manager:
        async with AsyncHeartbeat(action=fn, interval_s=5.0) as hb:
            await do_work()
    """

    __slots__ = (
        "_action", "_interval", "_cancel_token",
        "_task", "_disposed", "_unsub", "_stopped",
    )

    def __init__(
        self,
        action: Callable[[], Any],
        interval_s: float = 5.0,
        cancel_token: CancellationToken | None = None,
    ) -> None:
        self._action = action
        self._interval = interval_s
        self._cancel_token = cancel_token
        self._task: asyncio.Task | None = None  # type: ignore[type-arg]
        self._disposed = False
        self._unsub: Any = None
        self._stopped = asyncio.Event()

    async def start(self) -> None:
        """Inicia o heartbeat como asyncio.Task."""
        if self._task is not None or self._disposed:
            return
        if self._cancel_token is not None and self._cancel_token.is_cancelled:
            return

        # Cancelamento cross-thread via CancellationToken
        if self._cancel_token is not None:
            loop = asyncio.get_running_loop()

            def _on_cancel():
                loop.call_soon_threadsafe(self._stopped.set)

            self._unsub = self._cancel_token.on_cancelled(_on_cancel)

        self._task = asyncio.create_task(self._loop())

    def stop(self) -> None:
        """Para o heartbeat."""
        self._stopped.set()

    def dispose(self) -> None:
        """IDisposable — cancela task e limpa."""
        if self._disposed:
            return
        self._disposed = True
        self._stopped.set()
        if self._task is not None and not self._task.done():
            self._task.cancel()
        if self._unsub is not None:
            self._unsub.dispose()
            self._unsub = None

    async def __aenter__(self) -> AsyncHeartbeat:
        await self.start()
        return self

    async def __aexit__(self, *exc_info: Any) -> None:
        self.dispose()

    async def _loop(self) -> None:
        """Loop interno da task."""
        while not self._stopped.is_set():
            try:
                result = self._action()
                if asyncio.iscoroutine(result):
                    await result
            except Exception as exc:
                log.debug("Async heartbeat action falhou", error=str(exc))
            try:
                await asyncio.wait_for(self._stopped.wait(), timeout=self._interval)
                break  # stopped foi sinalizado
            except asyncio.TimeoutError:
                pass  # timeout = continuar loop

File: packages/gateway/src/heartbeat.ts
/**
 * Heartbeat generalizado para todos os canais.
 *
 * Problema real:
 *   O padrão "keep typing" existia hardcoded no Telegram (sendChatAction a cada 4s).
 *   WhatsApp, Discord e Slack não tinham feedback visual durante processamento.
 *
 * Solução:
 *   Heartbeat aceita qualquer callable como "ação de heartbeat":
 *   - Funciona com promises e sync alike
 *   - IDisposable — para automaticamente no dispose()
 *   - AbortSignal-aware — cancela se o processamento for abortado
 *
 * Uso:
 *   // Indicador de digitação durante processamento
 *   const hb = new ChannelHeartbeat({
 *     action: () => sendTypingIndicator(chatId),
 *     intervalMs: 4_000,
 *   });
 *   hb.start();
 *   try {
 *     const reply = await brain.prompt(text);
 *   } finally {
 *     hb.dispose();
 *   }
 *
 *   // Com AbortSignal (cancela se o cliente desconectar)
 *   const hb = new ChannelHeartbeat({
 *     action: () => sendTypingIndicator(chatId),
 *     intervalMs: 5_000,
 *     signal: req.signal,
 *   });
 */

import { childLogger } from "./logger.js";

const log = childLogger({ component: "heartbeat" });

// ---------------------------------------------------------------------------
// Tipos públicos
// ---------------------------------------------------------------------------

export interface HeartbeatOptions {
  /** Ação executada a cada intervalo. Pode ser async. */
  action: () => void | Promise<void>;
  /** Intervalo em ms entre execuções (default: 4000) */
  intervalMs?: number;
  /** AbortSignal para cancelamento externo */
  signal?: AbortSignal;
}

// ---------------------------------------------------------------------------
// ChannelHeartbeat
// ---------------------------------------------------------------------------

/**
 * Heartbeat generalizado — executa uma ação periodicamente enquanto vivo.
 * IDisposable: para no dispose() ou quando o AbortSignal for abortado.
 */
export class ChannelHeartbeat {
  private readonly action: () => void | Promise<void>;
  private readonly intervalMs: number;
  private readonly signal?: AbortSignal;

  private timer: ReturnType<typeof setInterval> | null = null;
  private disposed = false;
  private abortHandler: (() => void) | null = null;

  constructor(options: HeartbeatOptions) {
    this.action = options.action;
    this.intervalMs = options.intervalMs ?? 4_000;
    // exactOptionalPropertyTypes: só atribui se não for undefined
    if (options.signal !== undefined) {
      this.signal = options.signal;
    }
  }

  /** Inicia o heartbeat. Idempotente. */
  start(): this {
    if (this.disposed || this.timer !== null) return this;
    if (this.signal?.aborted) return this;

    // Registra handler de abort antes de iniciar
    if (this.signal) {
      this.abortHandler = () => this.dispose();
      this.signal.addEventListener("abort", this.abortHandler, { once: true });
    }

    this.timer = setInterval(() => {
      void this.tick();
    }, this.intervalMs);

    // Não bloqueia o processo
    if (this.timer.unref) this.timer.unref();

    return this;
  }

  /** Para o heartbeat. Idempotente. */
  stop(): void {
    if (this.timer !== null) {
      clearInterval(this.timer);
      this.timer = null;
    }
    if (this.abortHandler && this.signal) {
      this.signal.removeEventListener("abort", this.abortHandler);
      this.abortHandler = null;
    }
  }

  /** IDisposable — para e limpa. */
  dispose(): void {
    if (this.disposed) return;
    this.disposed = true;
    this.stop();
  }

  private async tick(): Promise<void> {
    if (this.disposed) return;
    try {
      await this.action();
    } catch (err) {
      log.debug({ err }, "Heartbeat action failed — ignorando");
    }
  }
}

// ---------------------------------------------------------------------------
// Helper: executa uma função com heartbeat automático
// ---------------------------------------------------------------------------

/**
 * Executa `fn` enquanto dispara `action` periodicamente como feedback visual.
 * O heartbeat para automaticamente quando `fn` termina (com sucesso ou erro).
 *
 * @example
 * const result = await withHeartbeat(
 *   () => sendTypingIndicator(chatId),
 *   5_000,
 *   () => brain.prompt(text),
 * );
 */
export async function withHeartbeat<T>(
  action: () => void | Promise<void>,
  intervalMs: number,
  fn: () => Promise<T>,
  signal?: AbortSignal,
): Promise<T> {
  const hb = new ChannelHeartbeat({ action, intervalMs, ...(signal !== undefined ? { signal } : {}) });
  hb.start();
  try {
    return await fn();
  } finally {
    hb.dispose();
  }
}
